Guilherme/engn 3811 modifiable topics

guilherme-brandao opened the pull request:
## Purpose of Changes and their Description
This PR introduces the ability for topic creators to update existing topics

- Add `UpdateTopic` message/handler so topic creators can update metadata and selected tuning parameters.
- Emit a `TopicUpdated` event whenever a topic change is applied.

## Link(s) to Ticket(s) or Issue(s) resolved by this PR
https://linear.app/alloralabs/issue/ENGN-3811/modifiable-topics

## Are these changes tested and documented?

- [x] If tested, please describe how. If not, why tests are not needed.
- [x] If documented, please describe where. If not, describe why docs are not needed.
- [x] Added to `Unreleased` section of `CHANGELOG.md`?

guilherme-brandao committed:
Add methods in msg server and keeper to update topic parameters

guilherme-brandao committed:
Apply topic updates during end block

guilherme-brandao committed:
Update methods to just edit specific properties

guilherme-brandao committed:
Fix remaining tests

guilherme-brandao committed:
Fix linter and tests

guilherme-brandao committed:
Update test to use new suite

guilherme-brandao committed:
Update functions to make just loss method and metadata updatable

xmariachi commented on x/emissions/types/errors.go:
```diff
@@ -97,4 +97,6 @@ var (
 	ErrMaxWhitelistInputArrayLengthExceeded      = errors.Register(ModuleName, 92, "max whitelist input array length exceeded")
 	ErrTopicAlreadyActive                        = errors.Register(ModuleName, 93, "topic already active")
 	ErrTopicCannotBeActivated                    = errors.Register(ModuleName, 94, "topic cannot be activated")
+	ErrNotTopicCreator                           = errors.Register(ModuleName, 95, "not topic creator")
```
Would modify it to sth like "ErrNotPermittedToModifyTopic" or similar, more aligned with other error names, and allows us to define what that means.

guilherme-brandao committed:
Fix lint

guilherme-brandao committed:
Add reviewer suggestion

guilherme-brandao committed:
Fix buf lint

spooktheducks commented on x/emissions/module/rewards/topic_rewards.go:
```diff
@@ -154,6 +154,26 @@ func GetAndUpdateActiveTopicWeights(
+		// Apply pending topic updates at epoch end, before calculating weight
+		hasPendingUpdate, err := k.HasPendingTopicUpdate(ctx, topicId)
+		if err != nil {
+			Logger(ctx).Warn("Error checking pending topic update", "topicId", topicId, "error", err)
```
Depending on the nature of the error, it seems that this could cause some non-determinism (i.e. problems with consensus re: app hash etc.).  Example: let's say one validator has a hardware error (disk full, whatever), and another does not.  They will come to different conclusions about the action to take here.

guilherme-brandao replied:
Yeah, I agree! It’s something we need to keep in mind for other parts of the code as well. Sometimes we use `continue` in loops and just log the error; sometimes we return the error; sometimes we panic. In cases where we need to iterate through topics, I think the old rule of thumb was: “if a topic has a problem, this shouldn’t affect other topics' processing.” I think we need to revisit that.

guilherme-brandao replied:
I think it's reasonable to return the errors in this case btw, I will change this.

spooktheducks commented on x/emissions/keeper/topic_weight.go:
```diff
@@ -102,11 +102,7 @@ func (k *Keeper) GetCurrentTopicWeight(
-	if err != nil {
-		return alloraMath.ZeroDec(), err
-	}
+func (k *Keeper) GetTopicWeightFromTopicId(ctx context.Context, topicId types.TopicId, params types.Params) (alloraMath.Dec, error) {
```
Is `params` used?

guilherme-brandao replied:
Yes, we use them inside when we call `GetCurrentTopicWeight`

guilherme-brandao committed:
Return error in topic rewards calc

guilherme-brandao committed:
fixes after merge

guilherme-brandao committed:
Use the most recent update tx

guilherme-brandao committed:
Enables immediate topic updates for inactive topics

guilherme-brandao committed:
Fix tests

guilherme-brandao committed:
Update changelog

xmariachi commented on x/emissions/keeper/msgserver/msg_server_topics.go:
```diff
@@ -97,3 +97,89 @@ func (ms msgServer) CreateNewTopic(ctx context.Context, msg *types.CreateNewTopi
+		return nil, errorsmod.Wrap(err, "Updated topic validation failed")
+	}
+
+	// If topic is inactive, apply immediately; for active topics, defer to epoch end.
```
Sending events: 
We need to send an event when topic gets updated. So, reputers can listen to the event of the change.
Also, the event needs to be triggered when the change comes into effect. So it would be sent within this tx, while the scheduled one will be triggered at end of epoch.

xmariachi commented on x/emissions/keeper/msgserver/msg_server_topics.go:
```diff
@@ -97,3 +97,89 @@ func (ms msgServer) CreateNewTopic(ctx context.Context, msg *types.CreateNewTopi
 	return &types.CreateNewTopicResponse{TopicId: topicId}, err
 }
+
+func (ms msgServer) UpdateTopic(ctx context.Context, msg *types.UpdateTopicRequest) (_ *types.UpdateTopicResponse, err error) {
```
I think this logic (which parameters get changed, how do they change, etc) should be in the keeper, not in the msgServer. And, the msgServer would just relay the call to the keeper.

xmariachi reviewed (commented):
More discussion to be had about the store, for now these two comments

guilherme-brandao committed:
Remove pending updates

guilherme-brandao committed:
Added new properties for update

guilherme-brandao committed:
Add new event for topic update

guilherme-brandao committed:
Move update logic to keeper

guilherme-brandao committed:
Remove redundant validation

xmariachi commented on CHANGELOG.md:
We should add the API changes (new tx) here, right? 
I'm thinking also that we should just state "API Changes" not "API Breaking Changes". wdyt? 
@spooktheducks @amimart

amimart replied:
I'm not sure it really brings a value to have a global section for all API changes, but it seems important to me to have a section dedicated to breaking changes. I'd add this change in the higher `### Added` section, and I guess the breaking changes `Added` section could be removed

guilherme-brandao replied:
I'd at least keep the API breaking changes separately. I think it's important have it.

xmariachi replied:
Yeah @amimart  @guilherme-brandao   `API Breaking Changes/Added` would never be API breaking because it would just be additive, so then it doesn't really makes sense - only `Changed` or `Removed` do.
So we could remove the `Added` subsection here.

amimart commented on x/emissions/keeper/keeper.go:
```diff
@@ -3253,6 +3253,105 @@ func (k *Keeper) TopicExists(ctx context.Context, topicId TopicId) (bool, error)
+
+	if !hasChanges {
+		return types.Topic{}, errorsmod.Wrap(types.ErrNoUpdateFields, "no fields to update")
+	}
```
Why erroring? We could simplify by letting it noop

guilherme-brandao replied:
I'm not sure about this behavior... It makes the experience ambiguous for the user. The transaction returns OK, but nothing is changed, the state stays the same. If the user accidentally sends an empty or invalid payload, it’s not obvious that nothing was actually applied.

xmariachi replied:
I think noop is the right call. The update is successful in that it leaves the topic in the state that the request is stating, so in that sense, it complies with the contract.
I don't think we should treat an empty payload differently. (They may accidentally do that, but what they requested has been "done": they requested noop, they got noop )

amimart replied:
> it leaves the topic in the state that the request is stating

Yes I agree, it successfully changes nothing, in the same way as a field present in the update msg that is the same as the value already present

amimart commented on x/emissions/keeper/keeper.go:
```diff
@@ -3253,6 +3253,105 @@ func (k *Keeper) TopicExists(ctx context.Context, topicId TopicId) (bool, error)
+
+	if err := msg.Validate(params.MaxStringLength); err != nil {
+		return types.Topic{}, err
+	}
```
I'd rather remove that validation as the topic is validated before being stored

guilherme-brandao replied:
good catch!

amimart commented on x/emissions/proto/emissions/v9/events.proto:
```diff
@@ -147,6 +147,10 @@ message EventCreateNewTopic {
 }
 
+message EventTopicUpdated {
+  uint64 topic_id = 1;
```
Should we add the topic as well? So that off chain nodes could directly use what's inside the event

amimart reviewed (commented):
Nice thx! Just left some suggestions

guilherme-brandao committed:
Update topic event

guilherme-brandao committed:
Remove extra validation

guilherme-brandao committed:
fix linter

guilherme-brandao committed:
Fix lint

amimart commented on x/emissions/proto/emissions/v9/events.proto:
```diff
@@ -147,6 +147,10 @@ message EventCreateNewTopic {
   emissions.v3.Topic topic = 1;
 }
 
+message EventTopicUpdated {
+  emissions.v3.Topic topic = 1;
+}
```
When subscribing to the websocket or searching txs by events, in order to be able to filter on an event's field it must be a root field. I think it'd be great to have at least the topic id as root field to allow filtering on it

guilherme-brandao committed:
Returns topic without error when no fields updated

Relates to ENGN-4064

guilherme-brandao committed:
Add topic id in the event

guilherme-brandao committed:
Change update topic to put style

guilherme-brandao committed:
Add autocli and remove unnecessary validation

amimart commented on x/emissions/keeper/keeper.go:
```diff
@@ -3253,6 +3253,66 @@ func (k *Keeper) TopicExists(ctx context.Context, topicId TopicId) (bool, error)
 }
 
+// UpdateTopic applies allowed changes to a topic.
+func (k *Keeper) UpdateTopic(ctx context.Context, msg *types.UpdateTopicRequest) (types.Topic, error) {
```
The message should be the responsibility of the msg server layer, the Keeper should stays independent from the message. Maybe we should pass the topic with its fields already updated directly

xmariachi replied:
Yep, `Request`s should not be in the keeper.

The `updatedTopic` can be built on the msg server, and being passed in to the `UpdateTopic(ctx context.Context, Topic updatedTopic)`  right?

amimart commented on x/emissions/keeper/keeper.go:
```diff
@@ -3253,6 +3253,66 @@ func (k *Keeper) TopicExists(ctx context.Context, topicId TopicId) (bool, error)
+	}
+
+	if topic.Creator != msg.Sender {
+		return types.Topic{}, types.ErrNotPermittedToModifyTopic
```
Maybe we can use the generic unauthorized error here:

```suggestion
		return types.Topic{}, sdkerrors.ErrUnauthorized
```

It's easier for clients to identify the underlying error when the code for a same error is consistent across modules

guilherme-brandao committed:
Fix keeper logic

guilherme-brandao committed:
Add reviewer suggestion

xmariachi commented on x/emissions/keeper/keeper.go:
```diff
@@ -3253,6 +3254,62 @@ func (k *Keeper) TopicExists(ctx context.Context, topicId TopicId) (bool, error)
+		return types.Topic{}, errorsmod.Wrap(err, "error getting params")
+	}
+
+	if topic.Creator != sender {
```
This authorization layer should be better done in the msg server, prior to calling the keeper.

xmariachi commented on x/emissions/keeper/keeper.go:
```diff
@@ -3253,6 +3254,62 @@ func (k *Keeper) TopicExists(ctx context.Context, topicId TopicId) (bool, error)
+	}
+
+	// Only allow updates to specific fields, and keep all other topic fields unchanged.
+	newTopic := topic
```
isn't `updatedTopic` already the topic to update? So, I think there wouldn't be a need to create a `newTopic`, no?.

xmariachi reviewed (changes requested):
Small changes requested on `UpdateTopic` fn

zale144 commented on x/emissions/proto/emissions/v9/tx.proto:
```diff
@@ -319,6 +321,31 @@ message CreateNewTopicResponse {
+
+  string sender = 1;
+  uint64 topic_id = 2;
+  string metadata = 3;
+  string loss_method = 4;
+  string alpha_regret = 5 [
+    (gogoproto.customtype) = "github.com/allora-network/allora-chain/math.Dec",
+    (gogoproto.nullable) = false
+  ];
+  string merit_sortition_alpha = 6 [
+    (gogoproto.customtype) = "github.com/allora-network/allora-chain/math.Dec",
+    (gogoproto.nullable) = false
+  ];
+  string p_norm = 7 [
+    (gogoproto.customtype) = "github.com/allora-network/allora-chain/math.Dec",
+    (gogoproto.nullable) = false
+  ];
```
maybe leave a comment on why only these fields are updated

xmariachi replied:
My 2ct here, I'm ot sure it is needed here in the proto. Each one has its own justification, so that could get too wordy. And also, justifications could be more meaningful into those topic fields that are _not_ modifiable - which are not defined here... so I'm ok not adding a comment here too.

xmariachi replied:
Sounds like it could be added to proper documentation, eg the `emissions` README. Currently it is pretty thin, could be used for rationales like these.

guilherme-brandao committed:
Fixes after review

zale144 commented on x/emissions/keeper/keeper.go:
```diff
@@ -3253,6 +3254,62 @@ func (k *Keeper) TopicExists(ctx context.Context, topicId TopicId) (bool, error)
+					return types.Topic{}, errorsmod.Wrap(err, "failed to check worker submission window")
+				}
+				if withinWindow {
+					return types.Topic{}, errorsmod.Wrap(types.ErrWorkerNonceWindowNotAvailable, "cannot update merit_sortition_alpha while worker window is open")
```
`loss_method`, `p_norm`, and `alpha_regret` are updated immediately. Since EMA-based regrets/weights carry across epochs, changing these mid-topic can mix losses with different semantics. should these not be epoch-boundary gated or versioned like `merit_sortition_alpha`?

zale144 replied:
@xmariachi any thoughts on this?

xmariachi replied:
They were considered to be safe, because they are being changed "for all actors at once". That is, the change (or not change) will impact every actor in an epoch - so it is fair. 
`merit_sortition_alpha` change could affect at tx time, so then this is why it is an exception.

guilherme-brandao committed:
Editable per topic c norm (#893)

## Purpose of Changes and their Description

- Move `c_norm` from global params to a per-topic field, updating proto
definitions, tx messages, keeper/msgserver logic, and synthesis/regret
calculations to read from `topic.CNorm`.
- Add migration v12 to drop global `c_norm` and copy the previous value
into every existing topic using a store iterator to avoid gaps.

## Link(s) to Ticket(s) or Issue(s) resolved by this PR
https://linear.app/alloralabs/issue/ENGN-4064/editable-per-topic-c-norm

## Are these changes tested and documented?

- [x] If tested, please describe how. If not, why tests are not needed.
- [x] If documented, please describe where. If not, describe why docs
are not needed.
- [x] Added to `Unreleased` section of `CHANGELOG.md`?

xmariachi commented on x/emissions/keeper/msgserver/msg_server_params.go:
```diff
@@ -126,9 +126,10 @@ func (ms msgServer) UpdateParams(ctx context.Context, msg *types.UpdateParamsReq
+	// NOTE: `c_norm` is deprecated as a global param and is now per-topic (Topic.CNorm).
+	// We keep the field in protobuf types for backwards compatibility, but ignore it here.
+	// If a client attempts to set it via UpdateParams, it will be a no-op.
+	_ = newParams.CNorm
```
if this is done just to pass linter: can we add a //nolint directive instead somehow?

guilherme-brandao replied:
It wasn’t for the linter, it’s just to make it explicit that c_norm is ignored. Just removed it!

guilherme-brandao committed:
Remove unnecessary code
